src/core_plugins/i18n/server/i18n/i18n.js

```diff
@@ -0,0 +1,123 @@
+};
+
+/**
+ * The translation file is registered with i18n plugin. The plugin contains a list of registered translation file paths per language.
+ * @param {string} absolutePluginTranslationFilePath - Absolute path to the translation file to register.
+ */
+export function registerTranslations(absolutePluginTranslationFilePath) {
+  const locale = getLocaleFromFileName(absolutePluginTranslationFilePath);
+
+  registeredTranslations[locale] = _.get(registeredTranslations, locale, []).concat(absolutePluginTranslationFilePath);
+};
+
+/**
+ * Unregister translation files
+ */
+export function unregisterTranslations() {
+  registeredTranslations = {};
+};
+
+/**
+ * Set the default locale
+ * @param {String} locale - Locale to set
+ */
+export function setDefaultLocale(locale) {
+  defaultLocale = locale;
+}
+
+function getTranslationsForLocale(locale) {
+  if (!registeredTranslations.hasOwnProperty(locale)) {
+    return Promise.resolve({});
+  }
+
+  const translationFiles = registeredTranslations[locale];
+  const translations = _.map(translationFiles, (filename) => {
+    return asyncReadFile(filename, 'utf8')
+    .then(fileContents => JSON.parse(fileContents))
+    .catch(SyntaxError, function (e) {
+      throw new Error('Invalid json in ' + filename);
+    })
+    .catch(function (e) {
+      throw new Error('Cannot read file ' + filename);
+    });
+  });
+
+  return Promise.all(translations)
+  .then(translations => _.assign({}, ...translations));
+};
+
+function getLocaleFromFileName(fullFileName) {
+  if (_.isEmpty(fullFileName)) throw new Error('Filename empty');
+
+  const fileExt = path.extname(fullFileName);
+  if (fileExt.length <= 0 || fileExt !== TRANSLATION_FILE_EXTENSION) {
+    throw new Error('Translations must be in a JSON file. File being registered is ' + fullFileName);
+  }
+
+  return path.basename(fullFileName, TRANSLATION_FILE_EXTENSION);
+}
+
+function getTranslationLocale(languageTags) {
+  const registeredLocales = getRegisteredTranslationLocales();
+  const tagLen = languageTags.length;
+  for (let indx = 0; indx < tagLen; indx++) {
+    const tag = languageTags[indx];
+    if (isLocaleExactMatch(tag, registeredLocales)) {
+      return tag;
+    }
+    const localeBestCaseMatch = getLocaleBestCaseMatch(tag, registeredLocales);
+    if (!_.isEmpty(localeBestCaseMatch)) {
+      return localeBestCaseMatch;
+    }
+  }
+  return '';
+}
+
+function isLocaleExactMatch(languageTag, registeredLocales) {
+  return  _.contains(registeredLocales, languageTag);
+}
+
+function getLocaleBestCaseMatch(languageTag, registeredLocales) {
+  // Find the first registered locale that begins with one of the language codes from the provided language tag.
+  // For example, if there is an 'en' language code, it would match an 'en-US' registered locale.
+  const languageCode = _.first(languageTag.split('-')) || [];
+  return _.find(registeredLocales, (locale) => _.startsWith(locale, languageCode));
+}
```

Comment:
Now that you're looping through the languageTags one at a time, I think you can wrap `isLocaleExactMatch` and `getLocaleBestCaseMatch` into one `getBestLocaleMatch` function which also allows you to simplify `getTranslationLocale`

```
function getTranslationLocale(languageTags) {
  let locale;
  _.forEach(languageTags, (tag) => {
      locale = locale || getBestLocaleMatch(tag);
  }
  return locale;
}

function getBestLocaleMatch(languageTag, registeredLocales) {
  if (_.contains(registeredLocales, languageTag) {
    return languageTag;
  }

  // Find the first registered locale that begins with one of the language codes from the provided language tag.
  // For example, if there is an 'en' language code, it would match an 'en-US' registered locale.
  const languageCode = _.first(languageTag.split('-')) || [];
  return _.find(registeredLocales, (locale) => _.startsWith(locale, languageCode));
}